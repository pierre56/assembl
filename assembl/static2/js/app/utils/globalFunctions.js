// @flow
import moment from 'moment';
import type Moment from 'moment';
import { type Map } from 'immutable';

import { getDisplayedPhaseIdentifier } from './timeline';
import { HARVESTABLE_PHASES, ICONS_PATH, PICTURE_BASE_URL, PICTURE_EXTENSION, PublicationStates } from '../constants';

const getInputValue = (id: string) => {
  const elem = document.getElementById(id);
  const value = elem instanceof HTMLInputElement ? elem.value : null;
  return value;
};

const getScriptText = (id: string) => {
  const elem = document.getElementById(id);
  const text = elem instanceof HTMLScriptElement ? elem.text : null;
  return text;
};

export const getDiscussionId = () => getInputValue('discussion-id');

export const getDiscussionSlug = () => getInputValue('discussion-slug');

export const encodeUserIdBase64 = (userId: string | null) => (userId ? btoa(`AgentProfile:${userId}`) : null);

// cache userId to avoid accessing the dom at each permission check
let userId;
export const getConnectedUserId = (base64: boolean = false) => {
  if (userId === undefined) {
    userId = getInputValue('user-id');
  }
  return base64 ? encodeUserIdBase64(userId) : userId;
};

export const getConnectedUserName = () => getInputValue('user-displayname');

export const isConnectedUser = (currentUserId: ?number) => String(currentUserId) === getConnectedUserId();

export const moveElementToFirstPosition = (array: Array<any>, targetElement: any) => {
  const others = array.filter(elelement => elelement !== targetElement);
  return [targetElement, ...others];
};

// cache permissions to avoid accessing the dom at each permission check
let permissions;
export const getConnectedUserPermissions = () => {
  if (permissions === undefined) {
    const scriptText = getScriptText('permissions-json') || '[]';
    permissions = JSON.parse(scriptText);
  }
  return permissions;
};

export function getAuthorizationToken<T>(location: { query: { token: T } }) {
  return 'token' in location.query ? location.query.token : null;
}

export const getProvidersData = () => {
  const data = getScriptText('login-providers');
  try {
    return data && JSON.parse(data); // $flowfixme
  } catch (e) {
    return null;
  }
};

export const getPossibleErrorMessage = () => {
  const errorMessageElem = document.getElementById('errorMessage');
  const data = errorMessageElem && errorMessageElem.innerHTML;
  return data;
};

export function getSortedArrayByKey<KeyType>(arr: Array<{ [KeyType]: number }>, key: KeyType) {
  arr.sort((a, b) => {
    if (a[key] < b[key]) {
      return -1;
    } else if (a[key] > b[key]) {
      return 1;
    }
    return 0;
  });
  return arr;
}

export const isDateExpired = (date1: number, date2: number) => date1 > date2;

export const getNumberOfDays = (date1: number, date2: number) => {
  const days = (date1 - date2) / (1000 * 60 * 60 * 24);
  return Math.round(days);
};

export const calculatePercentage = (value1: number, value2: number) => Math.round(value1 * 100 / value2 * 100) / 100;

/*
  Handrolled instead of using lodash
  Because lodash/capitalize lowercases everything else
*/
export const capitalize = (s: string) => s.charAt(0).toUpperCase() + s.slice(1);

export const getDocumentScrollTop = () =>
  window.pageYOffset ||
  (document.documentElement && document.documentElement.scrollTop) ||
  (document.body && document.body.scrollTop) ||
  0;

export const getDomElementOffset = (el: HTMLElement) => {
  const rect = el.getBoundingClientRect();
  const scrollTop = getDocumentScrollTop();
  const scrollLeft =
    window.pageXOffset ||
    (document.documentElement && document.documentElement.scrollLeft) ||
    (document.body && document.body.scrollLeft) ||
    0;
  return { top: rect.top + scrollTop, left: rect.left + scrollLeft };
};

export const computeDomElementOffset = (ref: HTMLElement, offset: { top?: number, left?: number }) => {
  // inspired from jquery.setOffset
  const curOffset = getDomElementOffset(ref);
  const curCSS = window.getComputedStyle(ref);
  const curTop = parseFloat(curCSS.top) || 0;
  const curLeft = parseFloat(curCSS.left) || 0;
  const result = { top: curTop, left: curLeft };

  if (typeof offset.top === 'number') result.top = offset.top - curOffset.top + curTop;
  if (typeof offset.left === 'number') result.left = offset.left - curOffset.left + curLeft;
  return result;
};

export const createEvent = (
  typeArg: string,
  eventInit: { bubbles: boolean, cancelable: boolean } = { bubbles: false, cancelable: false }
) => {
  // inspired from https://developer.mozilla.org/en-US/docs/Web/API/Event/Event
  const event = document.createEvent('Event'); // we can't use 'new Event()' because ie
  event.initEvent(typeArg, eventInit.bubbles, eventInit.cancelable);
  return event;
};

/*
  Get basename from a unix or windows path
*/
export const getBasename = (path: string) =>
  path
    .split('\\')
    .pop()
    .split('/')
    .pop();

export const hexToRgb = (c: string) => {
  if (!c) return '';
  const hex = c.replace(/[^0-9A-F]/gi, '');
  const bigint = parseInt(hex, 16);
  const r = (bigint >> 16) & 255; // eslint-disable-line
  const g = (bigint >> 8) & 255; // eslint-disable-line
  const b = bigint & 255; // eslint-disable-line

  return [r, g, b].join();
};

export const isMobile = {
  android: () => navigator.userAgent.match(/Android/i),
  blackberry: () => navigator.userAgent.match(/BlackBerry/i),
  ios: () => navigator.userAgent.match(/iPhone|iPad|iPod/i),
  opera: () => navigator.userAgent.match(/Opera Mini/i),
  windows: () => navigator.userAgent.match(/IEMobile/i),
  any: () => isMobile.android() || isMobile.blackberry() || isMobile.ios() || isMobile.opera() || isMobile.windows()
};

// works for SCREAMING_SNAKE_CASE, snake_case or cRazY_SNAKE_case
export const snakeToCamel = (string: string) => string.toLowerCase().replace(/_[a-z]/g, match => match[1].toUpperCase());

export const getCookieItem = (sKey: string) => {
  if (!sKey) {
    return null;
  }
  return (
    decodeURIComponent(
      document.cookie.replace(
        new RegExp(`(?:(?:^|.*;)\\s*${encodeURIComponent(sKey).replace(/[-.+*]/g, '\\$&')}\\s*\\=\\s*([^;]*).*$)|^.*$`),
        '$1'
      )
    ) || null
  );
};

const cookie = (key: string, value: string, expires: string, sensitive: boolean = false, path: string = '/'): string => {
  // As a security policy, we want all cookies that are set to be set according to the scheme
  // of the URL
  // In order to test sensitivity, utilize the debateData's requireSecureConnection and what type of
  // data is being put as a cookie
  let template = `${key}=${value};path=${path};expires=${expires}`;
  if (sensitive) {
    template += ';secure;HttpOnly;';
  }
  return template;
};

export function setCookieItem(name: string, value: any, sensitive: boolean = false) {
  const date = new Date();
  date.setMonth(date.getMonth() + 13);
  document.cookie = cookie(name, value, date.toString(), sensitive);
}

export function deleteCookieItem(name: string, path: string) {
  if (getCookieItem(name)) {
    // Set an expiration date in the past to delete the cookie in the browser
    const expiryDate = 'Thu, 28 Aug 1993 00:00:01 GTM';
    document.cookie = cookie(name, '', expiryDate, false, path);
  }
}

export const createRandomId = (): string => Math.round(Math.random() * -1000000).toString();

export const isOrContains = (n: any, c: any) => {
  const container = c;
  let node = n;
  while (node) {
    if (node === container) {
      return true;
    }
    node = node.parentNode;
  }
  return false;
};

export const elementContainsSelection = (el: any) => {
  let sel;
  if (window.getSelection) {
    sel = window.getSelection();
    if (sel.rangeCount > 0) {
      for (let i = 0; i < sel.rangeCount; i += 1) {
        if (!isOrContains(sel.getRangeAt(i).commonAncestorContainer, el)) {
          return false;
        }
      }
      return true;
    }
    // $FlowFixMe Property selection is missing in Document
  } else if (sel && sel === document.selection && sel.type !== 'Control') {
    return isOrContains(sel.createRange().parentElement(), el);
  }
  return false;
};

type ItemsById = Map<string, any>;
/* move item up (i.e. change its order) in a *ById state */
export const moveItemUp = (itemsById: ItemsById, id: string): ItemsById => {
  let newItemsById = itemsById;
  let itemsInOrder = itemsById
    .filter(item => !item.get('_toDelete'))
    .sortBy(item => item.get('order'))
    .map(item => item.get('id'))
    .toList();
  const idx = itemsInOrder.indexOf(id);
  itemsInOrder = itemsInOrder.delete(idx).insert(idx - 1, id);
  let order = 1;
  itemsInOrder.forEach((itemId) => {
    if (newItemsById.getIn([itemId, 'order']) !== order) {
      newItemsById = newItemsById.setIn([itemId, 'order'], order).setIn([itemId, '_hasChanged'], true);
    }
    order += 1;
  });

  return newItemsById;
};

/* move item down (i.e. change its order) in a *ById state */
export const moveItemDown = (itemsById: ItemsById, id: string): ItemsById => {
  let newItemsById = itemsById;
  let itemsInOrder = itemsById
    .filter(item => !item.get('_toDelete'))
    .sortBy(item => item.get('order'))
    .map(item => item.get('id'))
    .toList();
  const idx = itemsInOrder.indexOf(id);
  itemsInOrder = itemsInOrder.delete(idx).insert(idx + 1, id);
  let order = 1;
  itemsInOrder.forEach((itemId) => {
    if (newItemsById.getIn([itemId, 'order']) !== order) {
      newItemsById = newItemsById.setIn([itemId, 'order'], order).setIn([itemId, '_hasChanged'], true);
    }
    order += 1;
  });
  return newItemsById;
};

export function isHarvestable(params: RouterParams) {
  return HARVESTABLE_PHASES.includes(getDisplayedPhaseIdentifier(params));
}

export function fromGlobalId(id: string): string | null {
  return id ? atob(id).split(':')[1] : null;
}

export function getIconPath(icon: string, color: string = '') {
  return color ? `${ICONS_PATH}/${color}/${icon}` : `${ICONS_PATH}/${icon}`;
}

export const getPostPublicationState = (isDebateModerated: boolean, connectedUserIsAdmin: boolean): string => {
  if (!isDebateModerated || connectedUserIsAdmin) {
    return PublicationStates.PUBLISHED;
  }

  return PublicationStates.SUBMITTED_AWAITING_MODERATION;
};

// We `pictureId + 1` because there is no image in the S3 bucket with 0 as an id
export const getPictureUrl = (pictureId: number) => `${PICTURE_BASE_URL}${pictureId + 1}${PICTURE_EXTENSION}`;

export function compareByTextPosition(extractA: ?FictionExtractFragment, extractB: ?FictionExtractFragment) {
  // Sort extracts by position in body's paragraphs
  if (!!extractA && !!extractB) {
    const ATfi = extractA.textFragmentIdentifiers && extractA.textFragmentIdentifiers[0];
    const BTfi = extractB.textFragmentIdentifiers && extractB.textFragmentIdentifiers[0];
    const ATfiXPath = ATfi && ATfi.xpathStart;
    const BTfiXPath = BTfi && BTfi.xpathStart;

    const regex = /p\[([^)]+)\]/;
    const AMatchParagraph = ATfiXPath && ATfiXPath.match(regex);
    const BMatchParagraph = BTfiXPath && BTfiXPath.match(regex);

    if (!!AMatchParagraph && !!BMatchParagraph) {
      const AParagraph = parseInt(AMatchParagraph[1], 10);
      const BParagraph = parseInt(BMatchParagraph[1], 10);

      if (AParagraph > BParagraph) return 1;
      if (AParagraph === BParagraph) {
        // If extracts are in same paragraph, compare by offset
        const AOffsetStart = ATfi && ATfi.offsetStart;
        const BOffsetStart = BTfi && BTfi.offsetStart;
        if (!!AOffsetStart && !!BOffsetStart && AOffsetStart > BOffsetStart) return 1;
      }
    }
  }
  // For all other cases: (<), if extract null, not found in text => set first
  return -1;
}

export const convertISO8601StringToDate = (s: string): Moment => moment(s).utc();