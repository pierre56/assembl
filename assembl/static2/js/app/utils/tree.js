// @flow
import { PublicationStates } from '../constants';
import { connectedUserIsAdmin } from './permissions';
import { getConnectedUserId } from './globalFunctions';

type NodeType = {
  id: string,
  ancestors: Array<string>
};

type TreeNodeType = {
  id: string,
  parentId: string
};

type TreeType<T> = {
  roots: Array<T>,
  descendants: Array<T>
};

/**
 * @param {T: NodeType} The type of the nodes
 * @param {T} The root node.
 * @param {Array<T>} An array of nodes.
 * @returns {Array<T>} Returns the direct child nodes of the root.
 */
export function getChildren<T: NodeType>(rootNode: T, nodes: Array<T>): Array<T> {
  return nodes.filter(node => node.ancestors && node.ancestors.includes(rootNode.id));
}

/**
 * @param {T: NodeType} The type of the nodes
 * @param {Array<T>} An array of nodes.
 * @returns {{roots: Array<T>, descendants: Array<T>}} Returns the partial tree composed of all the root nodes and their children.
 */
export function getPartialTree<T: NodeType>(nodes: Array<T>): TreeType<T> {
  let ids = nodes.map(node => node.id);
  const roots = nodes.filter(node => !node.ancestors || node.ancestors.every(a => !ids.includes(a)));
  ids = roots.map(node => node.id);
  const descendants = nodes.filter(node => !ids.includes(node.id));
  return {
    roots: roots,
    descendants: descendants
  };
}

/**
 * @param {T: TreeNodeType} The type of the nodes
 * @param {Array<T>} An array of nodes.
 * @returns {{roots: Array<T>, descendants: Array<T>}} Returns the partial tree composed of all the root nodes and their children.
 */
export function getPartialTreeByParentId<T: TreeNodeType>(rootId: string, nodes: Array<T>): TreeType<T> {
  const result = {
    roots: [],
    descendants: []
  };
  if (nodes.length === 0) return result;
  const roots = rootId ? nodes.filter(item => !item.parentId || item.parentId === rootId) : nodes;
  if (roots.length === 0) return result;
  result.roots = roots;
  const rootsIds = roots.map(item => item.id);
  result.descendants = nodes.filter(item => !rootsIds.includes(item.id));
  return result;
}

/**
 * @param {T: TreeNodeType} The type of the nodes
 * @param {Array<T>} An array of nodes.
 * @returns {Array<T>} Returns the tree composed of all the root nodes and their children.
 */
export function getTree<T: TreeNodeType>(rootId: string, nodes: Array<T>, childrenName: string = 'children'): Array<T> {
  const { roots, descendants } = getPartialTreeByParentId(rootId, nodes);
  if (roots.length === 0) return [];
  return roots.map(item => ({ ...item, [childrenName]: getTree(item.id, descendants, childrenName) }));
}

/**
 * @param {T: TreeNodeType} The type of the nodes
 * @param {Array<T>} The item.
 * @param {Array<T>} An array of nodes.
 * @returns {Array<T>} Returns the ancestors included in nodes of the item.
 */
export function getAncestors<T: TreeNodeType>(item: T | null, nodes: Array<T>): Array<T> {
  const result = [];
  if (!item || !item.parentId) return result;
  const parent = nodes.find(node => item && node.id === item.parentId);
  if (parent) {
    result.push(parent);
    result.push(...getAncestors(parent, nodes));
  }
  return result;
}

/**
 * @param {T: TreeNodeType} The type of the nodes
 * @param {T | null} The item.
 * @param {Array<T>} An array of nodes.
 * @param {(items: Array<T>) => Array<T>} A sorter.
 * @returns {Array<T>} Returns the path of the item.
 */
export function getPath<T: TreeNodeType>(item: T, nodes: Array<T>, sortItems: (items: Array<T>) => Array<T>): Array<number> {
  const originalItem = nodes.find(n => n.id === item.id);
  if (!originalItem) return [];
  const ancestors: Array<T> = getAncestors(originalItem, nodes);
  const path = [];
  const nodesIds = nodes.map(n => n.id);
  const roots = nodes.filter(n => !nodesIds.includes(n.parentId));
  if (ancestors.length === 0) {
    if (roots.includes(originalItem)) {
      path.push(sortItems(roots).indexOf(originalItem));
    }
  } else {
    let node = originalItem;
    ancestors.forEach((ancestor) => {
      let children = nodes.filter(n => n.parentId === ancestor.id);
      children = sortItems ? sortItems(children) : children;
      path.push(children.indexOf(node));
      node = ancestor;
    });
    if (roots.includes(node)) {
      path.push(sortItems(roots).indexOf(node));
    }
  }
  return path.reverse();
}

export const isPostVisible = (postData: ChildType): boolean => {
  const connectedUserId = getConnectedUserId();
  const userIsAdmin = connectedUserIsAdmin();
  const userIsAuthor = connectedUserId === postData.creator.userId.toString();
  const isPublished = postData.publicationState === PublicationStates.PUBLISHED;
  const isPending = postData.publicationState === PublicationStates.SUBMITTED_AWAITING_MODERATION;
  return isPublished || (isPending && (userIsAdmin || userIsAuthor));
};