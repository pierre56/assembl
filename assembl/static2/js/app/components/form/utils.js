// @flow
import { EditorState } from 'draft-js';
import type { ApolloClient } from 'react-apollo';
import moment from 'moment';
import { I18n } from 'react-redux-i18n';
import flatMap from 'lodash/flatMap';

import type {
  I18nValue,
  FileValue,
  FileVariable,
  MutationsPromises,
  I18nRichTextValue,
  SaveStatus,
  DatePickerInput
} from './types.flow';
import { convertEditorStateToHTML, convertEntriesToEditorState, uploadNewAttachments } from '../../utils/draftjs';
import { displayAlert } from '../../utils/utilityManager';
import { runSerial } from '../administration/saveButton';
import UploadDocument from '../../graphql/mutations/uploadDocument.graphql';

export function i18nValueIsEmpty(v: I18nValue): boolean {
  return (
    !v ||
    Object.keys(v)
      .map(key => v[key]) // flow doesn't treat Object.values as expected, see: https://github.com/facebook/flow/issues/2221
      .every(s => s.length === 0)
  );
}

export function richTextI18nValueIsEmpty(v: I18nRichTextValue): boolean {
  return (
    !v ||
    Object.keys(v)
      .map(key => v[key]) // flow doesn't treat Object.values as expected, see: https://github.com/facebook/flow/issues/2221
      .every(es => !es || !es.getCurrentContent().hasText())
  );
}

// [{ localeCode: 'fr', value: 'foo' }] => { fr: 'foo' }
export function convertEntriesToI18nValue<T>(
  _entries: Array<{
    localeCode: string,
    value: T
  }>
): { [string]: T } {
  const entries = _entries || [];
  return entries.reduce((result, item) => {
    const { localeCode, value } = item;
    return {
      ...result,
      [localeCode]: value
    };
  }, {});
}

export function convertEntriesToI18nRichText(entries: RichTextLangstringEntries): I18nRichTextValue {
  return convertEntriesToI18nValue(convertEntriesToEditorState(entries));
}

export function convertISO8601StringToDateTime(_entry: string): DatePickerInput {
  if (_entry) {
    const t = moment(_entry, moment.ISO_8601).utc();
    if (t.isValid()) {
      return { time: t };
    }
  }
  return { time: null };
}

export function convertDateTimeToISO8601String(_entry: DatePickerInput): string | null {
  try {
    if (_entry && _entry.time) return _entry.time.utc().toISOString();
    return null;
  } catch (e) {
    return null;
  }
}

export function getValidationState(error: ?string, touched: ?boolean): ?string {
  return touched && error ? 'error' : null;
}

export const createSave = (successMsgId: string) => async (mutationsPromises: MutationsPromises): Promise<SaveStatus> => {
  let status = 'PENDING';
  await runSerial(mutationsPromises)
    .then(() => {
      status = 'OK';
      displayAlert('success', I18n.t(successMsgId));
    })
    .catch((error) => {
      status = 'KO';
      displayAlert('danger', error.message, false, 30000);
    });

  return status;
};

export function convertToEntries(valuesByLocale: I18nValue): LangstringEntries {
  return Object.keys(valuesByLocale).map(locale => ({
    localeCode: locale,
    value: valuesByLocale[locale]
  }));
}

type RichTextVariables = {
  attachments: Array<string>,
  entries: LangstringEntries
};

/*
  Upload new attachments and return variables for mutation
*/
export async function convertRichTextToVariables(
  valuesByLocale: I18nRichTextValue,
  client: ApolloClient
): Promise<RichTextVariables> {
  const uploadDocument = options => client.mutate({ mutation: UploadDocument, ...options });
  const results = Object.keys(valuesByLocale).map(async (locale) => {
    const result = await uploadNewAttachments(valuesByLocale[locale], uploadDocument);
    const { documentIds, contentState } = result;
    return {
      documentIds: documentIds,
      localeCode: locale,
      value: convertEditorStateToHTML(EditorState.createWithContent(contentState))
    };
  });

  const variables = await Promise.all(results);
  const attachments = flatMap(variables, v => v.documentIds);
  const entries = variables.map(v => ({
    localeCode: v.localeCode,
    value: v.value
  }));

  return {
    attachments: attachments,
    entries: entries
  };
}

export function getFileVariable(img: ?FileValue, initialImg: ?FileValue): FileVariable {
  if (initialImg && !img) {
    return 'TO_DELETE';
  }

  // If thematic.img.externalUrl is a File,
  // we need to send image: null if we didn't change the image.
  const variab = img && img.externalUrl instanceof File ? img.externalUrl : null;
  return variab;
}