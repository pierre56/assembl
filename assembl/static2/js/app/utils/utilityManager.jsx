import React from 'react';
import { I18n, Translate } from 'react-redux-i18n';
import { Link } from 'react-router';
import { getFullPath } from '../utils/routeMap';
import { getConnectedUserId } from '../utils/globalFunctions';
import SocialShare from '../components/common/socialShare';
import LoginButton from '../components/common/loginButton';

/*
  Singleton object that will contain the AlertManager, ModalManager which will
  be used to show/hide/manipulate the alert/modal system
*/
class UtilityManager {
  setComponent(comp) {
    this.component = comp;
  }
}

export const modalManager = new UtilityManager();

export const alertManager = new UtilityManager();

export const displayAlert = (style, msg, topPosition = false, time = 4000) => {
  /*
    alertStyle:String => bootstrap class (success, warning, danger or info)
    alertMsg:String => the message displayed in the alert
    showAlert:Boolean => to show/hide the alert
    topPosition:Boolean => if true, the top attribute should be 0
    time => the duration of the alert. If time === -1, it is persistent
  */
  alertManager.component.setState(
    // ensure that we hide the old alert if it was persistent
    {
      showAlert: false
    },
    () => {
      alertManager.component.setState({
        base: false,
        alertStyle: style,
        alertMsg: msg,
        showAlert: true,
        topPosition: topPosition
      });
    }
  );

  if (time !== -1) {
    setTimeout(() => {
      alertManager.component.setState({
        showAlert: false
      });
    }, time);
  }
};

export const displayModal = (title, body, footer, footerTxt, button = null, showModal = true, bsSize = null) => {
  /*
    title:String => the text in the header of the modal
    body:String => the text in the body of the modal
    footer:Boolean => to show/hide the footer of the modal
    footerTxt:String => the text in the footer of the modal,
    button:Object => the button in the footer of the modal
                    button.link:String => url of the button;
                    button.label:String => label of the button;
                    button.internalLink:Boolean => true if a Link from react-router is needed and false if a href is needed
  */
  if (document.activeElement) {
    document.activeElement.blur();
  }
  modalManager.component.setState({
    title: title,
    body: body,
    footer: footer,
    footerTxt: footerTxt,
    button: button,
    showModal: showModal,
    bsSize: bsSize,
    content: undefined
  });
};

export const displayCustomModal = (content, showModal = true, modalClass = null) => {
  /* display a modal with given content. You have to handle Modal.* components in your content */
  if (document.activeElement) {
    document.activeElement.blur();
  }
  modalManager.component.setState({
    content: content,
    showModal: showModal,
    modalClass: modalClass
  });
};

export const closeModal = () => {
  modalManager.component.setState({ showModal: false });
};

const getPathForModal = (type, params, elementId) => {
  switch (type) {
  case 'post':
    return { url: getFullPath('post', { ...params, element: elementId }) };
  case 'idea':
    return { url: getFullPath('idea', params) };
  case 'voteSession':
    return { url: getFullPath('voteSession', params) };
  case 'resourcesCenter':
    return { url: getFullPath('resourcesCenter', params) };
  case 'synthesis':
    return { url: getFullPath('synthesis', params) };
  case 'brightMirrorFiction':
    return { url: getFullPath('brightMirrorFiction', { ...params }) };
  default:
    return { url: getFullPath('post', { ...params, element: elementId }) };
  }
};

export const openShareModal = (options) => {
  const { type, title, routerParams, elementId, social, isFooter, footer } = options;
  const pathForModal = getPathForModal(type, routerParams, elementId);
  const modalBody = <SocialShare url={pathForModal.url} onClose={closeModal} social={social} />;
  return displayModal(title, modalBody, isFooter, footer);
};

export const inviteUserToLogin = () => {
  const body = (
    <div>
      <p>
        <Translate value="login.loginModalBody" />
      </p>
      <br />
      <LoginButton label={I18n.t('login.loginModalFooter')} />
    </div>
  );
  displayModal(null, body, true, null, null, true);
};

/* if user is not connected, ask for login, else, execute given action */
export const promptForLoginOr = action => () => {
  const isUserConnected = getConnectedUserId(); // TO DO put isUserConnected in the store
  if (!isUserConnected) {
    inviteUserToLogin();
  } else {
    action();
  }
};

export const defaultAnchorAttributes = {
  rel: 'noopener no-referrer',
  target: '_blank'
};

/*
  An HOC that can be used when unclear whether to use Link or a tag (internal react-router
  or external url)
  @params AnchorComponent   Component : The component within the link
  @params urlData           Object : An object that has 'url' and 'local', describing whether link is internal or not
  @params anchorAttributes  Object: All attributes desired to be passed to the Link or a tag
  @params props             Object: All props desired to be passed to the AnchorComponent
*/
export const localAwareLink = AnchorComponent => ({ urlData, anchorAttributes, props }) => {
  const attrs = anchorAttributes || null;
  const componentProps = props || null;
  if (!urlData.local) {
    return (
      <a href={urlData.url} {...attrs}>
        <AnchorComponent {...componentProps} />
      </a>
    );
  }
  return (
    <Link to={urlData.url} {...attrs}>
      <AnchorComponent {...componentProps} />
    </Link>
  );
};