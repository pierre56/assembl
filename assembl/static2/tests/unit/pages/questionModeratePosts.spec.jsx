import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16.3';
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import QuestionModeratePosts from '../../../js/app/pages/questionModeratePosts';
import { connectedUserIsAdmin } from '../../../js/app/utils/permissions';
import { get, goTo } from '../../../js/app/utils/routeMap';

configure({ adapter: new Adapter() });

jest.mock('../../../js/app/utils/permissions');
jest.mock('../../../js/app/utils/routeMap');

describe('QuestionModeratePosts page', () => {
  const props = {
    imgUrl: 'https://foo.bar/umgurl',
    isModerating: true,
    timeline: [
      {
        id: 'phaseFoo',
        title: 'Survey',
        identifier: 'survey',
        start: '2018-09-27T20:00:00+00:00',
        end: '2100-01-27T20:00:00+00:00'
      }
    ],
    title: 'Foo',
    thematicTitle: 'Bar',
    thematicId: 'ThematicId',
    numPosts: 2,
    numContributors: 2,
    totalSentiments: 1,
    params: {
      questionIndex: 'Index',
      questionId: 'FooInd',
      slug: 'FooSlug'
    },
    phaseId: 'phaseFoo'
  };

  describe('componentDidMount method', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    it('should redirect to unauthorized if connected user is not admin', () => {
      connectedUserIsAdmin.mockImplementation(() => false);
      shallow(<QuestionModeratePosts {...props} />);
      expect(get).toHaveBeenCalledWith('unauthorizedAdministration', { slug: 'FooSlug' });
      expect(goTo).toHaveBeenCalled();
    });

    it('should not redirect to unauthorized if connected user is admin', () => {
      connectedUserIsAdmin.mockImplementation(() => true);
      shallow(<QuestionModeratePosts {...props} />);
      expect(get).not.toHaveBeenCalled();
      expect(goTo).not.toHaveBeenCalled();
    });
  });

  it('should display a question', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<QuestionModeratePosts {...props} />);
    const rendered = renderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });
});