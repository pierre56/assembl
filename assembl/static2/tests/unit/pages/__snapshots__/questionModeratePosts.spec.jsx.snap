// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionModeratePosts page should display a question 1`] = `
<Connect(Apollo(Component))
  imgUrl="https://foo.bar/umgurl"
  isModerating={true}
  numContributors={2}
  numPosts={2}
  params={
    Object {
      "questionId": "FooInd",
      "questionIndex": "Index",
      "slug": "FooSlug",
    }
  }
  phaseId="phaseFoo"
  thematicId="ThematicId"
  thematicTitle="Bar"
  timeline={
    Array [
      Object {
        "end": "2100-01-27T20:00:00+00:00",
        "id": "phaseFoo",
        "identifier": "survey",
        "start": "2018-09-27T20:00:00+00:00",
        "title": "Survey",
      },
    ]
  }
  title="Foo"
  totalSentiments={1}
/>
`;
